# Clean your Data With NumPy and Pandas

In [1]:
#Import Library
import numpy as np
import pandas as pd

In [2]:
# Read Data from your datasets folder
data = pd.read_csv('../datasets/clean-data/Book.csv')

In [3]:
# Create DataFrame from the CSV file
data.head() # To see First Five entery with column names

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

# Data exploration and Basic Hygiene

In [4]:
# Print Missing value
print(data.isnull().sum())

Identifier                   0
Edition Statement         7514
Place of Publication         0
Date of Publication        181
Publisher                 4195
Title                        0
Author                    1778
Contributors                 0
Corporate Author          8287
Corporate Contributors    8287
Former owner              8286
Engraver                  8287
Issuance type                0
Flickr URL                   0
Shelfmarks                   0
dtype: int64


In [5]:
# Drop columns with any missing values 
c=data.dropna(axis='columns')
c.head()

Identifier      Place of Publication  \
0         206                    London   
1         216  London; Virtue & Yorston   
2         218                    London   
3         472                    London   
4         480                    London   

                                               Title  \
0                  Walter Forbes. [A novel.] By A. A   
1  All for Greed. [A novel. The dedication signed...   
2  Love the Avenger. By the author of “All for Gr...   
3  Welsh Sketches, chiefly ecclesiastical, to the...   
4  [The World in which I live, and my place in it...   

                                   Contributors Issuance type  \
0                               FORBES, Walter.   monographic   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
3                   Appleyard, Ernest Silvanus.   monographic   
4                           BROOME, John Henry.   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [6]:
# Missing value is not Found as it is removed
print(c.isnull().sum())

Identifier              0
Place of Publication    0
Title                   0
Contributors            0
Issuance type           0
Flickr URL              0
Shelfmarks              0
dtype: int64


In [7]:
# Drop the rows where at least one element is missing.
data.dropna()


Empty DataFrame
Columns: [Identifier, Edition Statement, Place of Publication, Date of Publication, Publisher, Title, Author, Contributors, Corporate Author, Corporate Contributors, Former owner, Engraver, Issuance type, Flickr URL, Shelfmarks]
Index: []

Link For Data operation using pandas 
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html

# We can remove columns by passing columns names directly to columns parameters

### data.drop(columns=drop_col, inplace=True)
### data.head()

In [8]:
# drop unnessary columns 
# inplace and axis make change to object
# axis =1 drop column wise axis =0 for row wise
drop_col = ['Edition Statement',
           'Corporate Author',
           'Corporate Contributors',
           'Former owner',
           'Engraver',
           'Contributors',
           'Issuance type',
           'Shelfmarks']

data.drop(drop_col, inplace = True, axis = 1)
data.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                         Flickr URL  
0      A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
1  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
2  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
3  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...  
4  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...

In [9]:
# Missing value is not Found as it is removed
print(data.isnull().sum())

Identifier                 0
Place of Publication       0
Date of Publication      181
Publisher               4195
Title                      0
Author                  1778
Flickr URL                 0
dtype: int64


# Change Index of Dataframe

In [10]:
# Check unique values in  Columns
data['Identifier'].is_unique

True

In [11]:
data['Place of Publication'].is_unique

False

In [12]:
data['Date of Publication'].is_unique

False

In [13]:
data['Publisher'].is_unique

False

In [14]:
data['Title'].is_unique

False

In [15]:
data['Author'].is_unique

False

In [16]:
data['Flickr URL'].is_unique

True

In [17]:
# Replace the existing index with column name with unique value using set_index :
# To make change directly to object
data.set_index('Identifier', inplace=True)
data.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

### Replace the existing index with column name with unique value using set_index :
### data = data.set_index('Identifier')
### data.head()

In [18]:
# Access each record with loc[]
data.loc[480]

Place of Publication                                               London
Date of Publication                                                  1857
Publisher                                            Wertheim & Macintosh
Title                   [The World in which I live, and my place in it...
Author                                                          A., E. S.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 480, dtype: object

In [19]:
data.iloc[480]

Place of Publication                                               Oxford
Date of Publication                                                  1846
Publisher                                               John Henry Parker
Title                   The Baptistery ... By the author of “The Cathe...
Author                                                                NaN
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 192427, dtype: object

In [20]:
data.iloc[5]

Place of Publication                                               London
Date of Publication                                                  1875
Publisher                                               William Macintosh
Title                   [The World in which I live, and my place in it...
Author                                                          A., E. S.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 481, dtype: object


    We can see from above with
    loc[] => help in accessing data based on index
       whereas
    iloc[] => help in accessing data based on position
    index [480] data is same as in position [5]
   

# Clean Fields in the Data

In [21]:
# All data types are in object dtype
data.get_dtype_counts()

object    6
dtype: int64

In [22]:
data.loc[1905:, 'Date of Publication'].head(10)

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
2957           1873
3017           1866
3131           1899
4598           1814
4884           1820
Name: Date of Publication, dtype: object

<b>As stated above some book have multiple date of publication. Book can have only one date of publication.</b>

<b>We need to enforce a numeric value for date of publication to do calculations as:</b>
<li>Remove square brackets, like present: 1879 [1878]</li>
<li>Convert date ranges to their “start date”, wherever present: 1860-63; 1839, 38-54</li>
<li>Remove the dates like: [1897?] replace them with NumPy’s NaN : </li>
<li>Convert the string NaN to NumPy’s NaN value</li>

In [23]:
ex = data['Date of Publication'].str.extract(r'^(\d{4})', expand=False)
ex.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

In [24]:
# Convert to numerical values
data['Date of Publication'] = pd.to_numeric(ex)
data['Date of Publication'].dtypes

dtype('float64')

In [25]:
data['Date of Publication'].isnull().sum() / len(data)

0.11717147339205986

# Use str Methods with NumPy to clean Columns

<b>To clean Place of Publication we use str method with NumPy</b>

In [26]:
data['Place of Publication'].head()

Identifier
206                      London
216    London; Virtue & Yorston
218                      London
472                      London
480                      London
Name: Place of Publication, dtype: object

In [27]:
data.loc[data['Place of Publication']== 'London']

Place of Publication  Date of Publication                Publisher  \
Identifier                                                                      
206                      London               1879.0         S. Tinsley & Co.   
218                      London               1869.0    Bradbury, Evans & Co.   
472                      London               1851.0            James Darling   
480                      London               1857.0     Wertheim & Macintosh   
481                      London               1875.0        William Macintosh   
519                      London               1872.0               The Author   
1143                     London               1679.0                      NaN   
2854                     London               1865.0           E. Moxon & Co.   
4884                     London               1820.0        J. Hatchard & Son   
5382                     London               1847.0             Punch Office   
5385                     London                  NaN    Bradbury, Agnew & Co.   
5389                     London                  NaN    Bradbury, Agnew & Co.   
6036                     London               1805.0          C. & R. Baldwin   
8435                     London               1892.0   Effingham Wilson & Co.   
8440                     London               1894.0         Effingham Wilson   
13074                    London               1899.0         W. Thacker & Co.   
13364                    London                  NaN   Griffith, Farran & Co.   
14466                    London                  NaN             Adams & King   
15141                    London               1893.0          T. Fisher Unwin   
15146                    London               1894.0          T. Fisher Unwin   
16543                    London               1816.0              John Murray   
17602                    London               1897.0                J. Bowden   
18173                    London               1804.0            Vernor & Hood   
20960                    London               1853.0              Addey & Co.   
21119                    London               1874.0           Chapman & Hall   
21430                    London               1790.0               J. Johnson   
21782                    London                  NaN              J. Johnston   
22105                    London               1795.0               J. Debrett   
23565                    London               1877.0          Chatto & Windus   
24878                    London               1886.0       Hodder & Stoughton   
...                         ...                  ...                      ...   
3998513                  London               1893.0          Macmillan & Co.   
3998658                  London               1843.0      J. B. Nichols & Son   
3999230                  London               1813.0                      NaN   
4001224                  London               1877.0                      NaN   
4001226                  London               1868.0                      NaN   
4001363                  London               1898.0       A. Constable & Co.   
4002984                  London               1858.0                      NaN   
4006496                  London               1898.0             W. Heinemann   
4006497                  London               1894.0             W. Heinemann   
4006498                  London               1896.0             W. Heinemann   
4006577                  London               1832.0                      NaN   
4015049                  London               1889.0            Blackie & Son   
4027249                  London               1803.0                      NaN   
4029050                  London               1899.0              T. F. Unwin   
4029064                  London               1869.0                      NaN   
4030632                  London               1831.0                      NaN   
4041055                  London           

In [28]:
data.loc[data['Place of Publication']== 'Oxford']

Place of Publication  Date of Publication  \
Identifier                                             
111455                   Oxford                  NaN   
192427                   Oxford               1846.0   
262687                   Oxford               1887.0   
686597                   Oxford               1818.0   
962896                   Oxford               1892.0   
991105                   Oxford               1897.0   
991106                   Oxford               1899.0   
1211883                  Oxford               1895.0   
1259903                  Oxford               1895.0   
1341190                  Oxford               1892.0   
1553618                  Oxford               1640.0   
1565531                  Oxford               1893.0   
1565534                  Oxford               1893.0   
1630483                  Oxford               1885.0   
1635714                  Oxford               1829.0   
1635730                  Oxford               1829.0   
1664649                  Oxford               1824.0   
1664652                  Oxford               1846.0   
1688742                  Oxford               1858.0   
1781291                  Oxford               1826.0   
1781292                  Oxford               1827.0   
1818869                  Oxford               1663.0   
1887400                  Oxford               1888.0   
2102452                  Oxford               1889.0   
2202381                  Oxford               1844.0   
2299204                  Oxford               1658.0   
2502001                  Oxford               1893.0   
2539667                  Oxford                  NaN   
2573707                  Oxford               1830.0   
2698099                  Oxford               1892.0   
2714138                  Oxford               1823.0   
2740260                  Oxford               1847.0   
2740310                  Oxford               1859.0   
2740334                  Oxford               1824.0   
2740636                  Oxford               1818.0   
2740891                  Oxford               1778.0   
2779568                  Oxford               1882.0   
3036588                  Oxford               1892.0   
3039451                  Oxford               1875.0   
3354959                  Oxford               1766.0   
3519690                  Oxford               1893.0   
3557266                  Oxford               1892.0   
3605102                  Oxford                  NaN   
3999294                  Oxford               1885.0   

                                           Publisher  \
Identifier                                             
111455                              J. & J. Fletcher   
192427                             John Henry Parker   
262687                               Rowbottom & Son   
686597                              [Clarendon Press   
962896                                           NaN   
991105                               Clarendon Press   
991106                               Clarendon Press   
1211883                              B. H. Blackwell   
1259903                              Clarendon Press   
1341190                              Clarendon Press   
1553618                                    W. Turner   
1565531                              Clarendon Press   
1565534                              Clarendon Press   
1630483                                          NaN   
1635714                                D. A. Talboys   
1635730                                          NaN   
1664649                                          NaN   
1664652                                          NaN   
1688742                                          NaN   
1781291                                          NaN   
1781292                                          NaN   
1818869                         H. Hall for R. Davis   
1887400                              Clarendon Press   
2102452                           

<b>We see some row with Place of Publication as 'London' and 'Oxford' have unnecessary information.</b>

<b>Clean this column in one go use str.contains() to get a boolean mask.</b>

In [29]:
london = data['Place of Publication'].str.contains('London')
london.head()

Identifier
206    True
216    True
218    True
472    True
480    True
Name: Place of Publication, dtype: bool

In [30]:
oxford = data['Place of Publication'].str.contains('Oxford')
oxford.head()

Identifier
206    False
216    False
218    False
472    False
480    False
Name: Place of Publication, dtype: bool

In [31]:
# Combine all with NumPy
data['Place of Publication'] = np.where(london, 'London',
                                      np.where(oxford, 'Oxford',
                                               data['Place of Publication'].str.replace('-', ' ')))
data.head()

Place of Publication  Date of Publication              Publisher  \
Identifier                                                                    
206                      London               1879.0       S. Tinsley & Co.   
216                      London               1868.0           Virtue & Co.   
218                      London               1869.0  Bradbury, Evans & Co.   
472                      London               1851.0          James Darling   
480                      London               1857.0   Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

<b>There is some dirt in other columns too</b><br />
<b>To clean the dirt from entire dataset we use 'applymap' method</b>

# Next For 'applymap' method